openstack_controller/datadog_checks/openstack_controller/openstack_controller.py

```diff
@@ -358,74 +331,82 @@ def get_all_servers(self, tenant_to_name, instance_name):
+        # For each updated servers, we update the servers cache accordingly
+        for updated_server in updated_servers:
+            updated_server_status = updated_server.get('status')
+            updated_server_id = updated_server.get('id')
 
-        self.changes_since_time[instance_name] = datetime.utcnow().isoformat()
-
-        for server in servers:
-            new_server = dict()
-            new_server['server_id'] = server.get('id')
-            new_server['state'] = server.get('status')
-            new_server['server_name'] = server.get('name')
-            new_server['hypervisor_hostname'] = server.get('OS-EXT-SRV-ATTR:hypervisor_hostname')
-            new_server['tenant_id'] = server.get('tenant_id')
-            new_server['availability_zone'] = server.get('OS-EXT-AZ:availability_zone')
-
-            # Confirm that the new server has all the required fields
-            if not all(new_server[key] is not None for key in SERVER_FIELDS_REQ):
-                self.log.debug(
-                    "Server {} has None for one of the required keys."
-                    "Not collecting server metrics for this server."
-                    .format(new_server)
-                )
-                continue
-
-            # Update our cached list of servers
-            if (new_server['server_id'] not in self.server_details_by_id and
-                    new_server['state'] in DIAGNOSTICABLE_STATES):
-                self.log.debug("Adding server to cache: %s", new_server)
-                # The project may not exist if the server isn't in an active state
-                # Query for the project name here to avoid 404s
-                # TODO: why should we avoid 404? Let's just skip it until the next run
-                # If the project name conversion fails, we won't add this server
-
-                # We should not fail on 404 if it happen and we should adjust the cache on the next run if need be.
-                # In all cases we should avoid, making unecessary api calls
-                project_name = tenant_to_name.get(new_server['tenant_id'])
-                new_server['project_name'] = project_name
-                if new_server['project_name']:
-                    self.server_details_by_id[new_server['server_id']] = new_server
-            elif new_server['server_id'] in self.server_details_by_id and new_server['state'] in REMOVED_STATES:
-                self.log.debug("Removing server from cache: %s", new_server)
-                try:
-                    del self.server_details_by_id[new_server['server_id']]
-                except KeyError:
-                    self.log.debug("Server: %s has already been removed from the cache", new_server['server_id'])
-
-    def collect_server_metrics(self, server_details, tags=None, use_shortname=False):
+            if updated_server_status == 'ACTIVE':
+                # Add or update the cache
+                if tenant_to_name[updated_server.get('tenant_id')]:
```

Comment:
Shouldn't we always update the cache if a new metric came in with an ACTIVE state? Any reason to check it exists in this mapping?

Reply:
Or do we avoid this entirely if we don't have a project associated with the server?

Reply:
That is what the previous logic was doing. This logic specific logic even has a unit test `test_get_all_servers_with_project_name_none`. I am assuming it is legit

Reply:
Ok, yeah I think we actually discussed this in the past already. Makes sense 👍